In [3]:
import sys
import os
import pandas as pd
import datetime

sys.path.append(os.path.abspath(os.path.join('..')))

from src.data_loader import fetch_data_tiingo

# Config block
START_DATE = "2009-01-01"
END_DATE = datetime.datetime.today().strftime('%Y-%m-%d')
API_KEY = "08ec876fa0ec9c87cafff1fd07c6ffd95036203a"
ticker = "XLE"
data_path = f"../data/{ticker.lower()}_{START_DATE}_{END_DATE}.csv"

if os.path.exists(data_path):
    df = pd.read_csv(data_path, index_col=0, parse_dates=True)
else:
    df = fetch_data_tiingo(ticker, api_key=API_KEY, start_date=START_DATE, end_date=END_DATE)
    
    #Save for local use
    os.makedirs("../data", exist_ok=True)
    df.to_csv(data_path)

In [1]:
import sys
import os
import pandas as pd

# Allow importing from ../src
sys.path.append(os.path.abspath(os.path.join("..")))

from src.strategy import sma_crossover_strategy
from src.backtester import backtest
from src.metrics import sharpe_ratio, max_drawdown
from src.utils import plot_equity_curve

# Load the saved data
df = pd.read_csv("../data/xle_2009-01-01_2025-07-28.csv", index_col=0, parse_dates=True)

df_raw = df.copy()

# Apply strategy and run backtest
df = sma_crossover_strategy(df, short_window=20, long_window=50)
results = backtest(df)

# Compute metrics
sr = sharpe_ratio(results['Strategy_Returns'].dropna())
dd = max_drawdown(results['Equity_Curve'].dropna())


print(f"Sharpe Ratio: {sr:.2f}")
print(f"Max Drawdown: {dd:.2%}")

# Plot equity curve
plot_equity_curve(results)

Sharpe Ratio: -0.31
Max Drawdown: -89.45%


We applied a simple 20/50-day Simple Moving Average (SMA) crossover strategy to XLE, the Energy Select Sector SPDR ETF, covering the period from 2009 to July 2025. This strategy generates long-only signals when the short-term SMA (20-day) crosses above the long-term SMA (50-day).

Key performance metrics:
Sharpe Ratio: –0.31
Indicates that the strategy delivered negative risk-adjusted returns. It underperformed even a zero-return (risk-free) benchmark, on a volatility-adjusted basis.

Max Drawdown: –89.45%
The strategy suffered a near-complete loss of capital at its worst point, highlighting poor downside protection and high vulnerability to market noise.

Equity curve analysis:
The equity curve shows a steady and persistent drawdown from 2009 onward, with only short-lived recoveries. There are multiple failed breakouts and whipsaw periods where the SMA crossover generated false signals, resulting in frequent losses and erosion of capital.

Notably, there is no sustained upward trend in cumulative returns. Sharp drops (e.g., during 2015 oil crash, 2020 COVID, 2022 energy volatility) caused deep drawdowns. The curve shows volatility without reward a hallmark of a strategy out of sync with market structure.

Takeaways:
An obvious point, not all strategies generalise across all asset classes. SMA crossovers work better on trending assets (e.g., SPY, NASDAQ, BTC) than on cyclicals or commodities.
Backtesting is Discovery: Negative results are not failures. They help us reject poor hypotheses and avoid overfitting.
Comparing across assets, adding filters (e.g., volatility regimes), or hybrid signals may improve strategy viability.

Let us put to test the third takeaway. 

Adding a volatility filter: Don't trade below unless volatility is below a certain threshold (to avoid whipsaws).

Idea: Use Average True Range (ATR) or rolling standard deviation as a proxy for volatility. Only enter trades when volatility is below a defined percentile (eg: 50th percentile). 

Intuition:
High volatility = unstable signals → avoid
Low volatility = more stable regime → allow SMA crossover to execute